design/plankton/CommunicationLayerDesign.md

```diff
@@ -36,13 +36,161 @@ The IBEX team at ISIS, for example, have expressed interest in using Plankton. N
 
 ## Design Options
 
-### Option A
+### Adapter-Binder Approach
 
-Detail design option A here.
+Overview:
+
+* `Adapters` are bespoke per `Device` as well as per `Protocol`.
+* We term the `Hardware` as per the manufacturer specification to be `Device` + `Adapter`.
+* There is no separate binding.
+* There is no general concrete `Adapter` for any protocol. 
+* The process loop is kicked off by the `Environment`
+
+#### Adapters
+
+The problem with the current implementation is that the `Adapter` concept is too general and the `Binder` mechanism is too limited (1:1) to make linkup to a `Device` powerful or flexible enough.
+
+The `Adapters` in this design are per-protocol & per-device. So they can encapsulate all the logic necessary to drive the specific device that they wrap. 
+
+```python
+class StreamAdapterLinkum95(StreamAdapter):
```

Comment:
It depends a bit, I think for the EPICS case the base class can be quite "powerful" to make it easy to expose device properties as PVs. For Stream that might be different and the abstraction is "further up", but still it should be possible to eliminate some boiler plate.

Regarding the `ArgParser`-thing, probably the adapters have their own `ArgParser`, like they do now.